nixos/modules/services/networking/headscale.nix

```diff
@@ -47,204 +55,247 @@ in {
         '';
       };
 
-      address = mkOption {
-        type = types.str;
+      address = lib.mkOption {
+        type = lib.types.str;
         default = "127.0.0.1";
         description = ''
           Listening address of headscale.
         '';
         example = "0.0.0.0";
       };
 
-      port = mkOption {
-        type = types.port;
+      port = lib.mkOption {
+        type = lib.types.port;
         default = 8080;
         description = ''
           Listening port of headscale.
         '';
         example = 443;
       };
 
-      settings = mkOption {
+      settings = lib.mkOption {
         description = ''
           Overrides to {file}`config.yaml` as a Nix attribute set.
           Check the [example config](https://github.com/juanfont/headscale/blob/main/config-example.yaml)
           for possible options.
         '';
-        type = types.submodule {
+        type = lib.types.submodule {
           freeformType = settingsFormat.type;
 
+          imports = with lib; [
+            (mkAliasOptionModule ["acl_policy_path"] ["policy" "path"])
+            (mkAliasOptionModule ["db_host"] ["database" "postgres" "host"])
+            (mkAliasOptionModule ["db_name"] ["database" "postgres" "name"])
+            (mkAliasOptionModule ["db_password_file"] ["database" "postgres" "password_file"])
+            (mkAliasOptionModule ["db_path"] ["database" "sqlite" "path"])
+            (mkAliasOptionModule ["db_port"] ["database" "postgres" "port"])
+            (mkAliasOptionModule ["db_type"] ["database" "type"])
+            (mkAliasOptionModule ["db_user"] ["database" "postgres" "user"])
+            (mkAliasOptionModule ["dns_config" "base_domain"] ["dns" "base_domain"])
+            (mkAliasOptionModule ["dns_config" "domains"] ["dns" "search_domains"])
+            (mkAliasOptionModule ["dns_config" "magic_dns"] ["dns" "magic_dns"])
+            (mkAliasOptionModule ["dns_config" "nameservers"] ["dns" "nameservers" "global"])
```

Comment:
Why not
```suggestion
            (mkRenamedOptionModule ["dns_config" "nameservers"] ["dns" "nameservers" "global"])
```
?

Reply:
Actually I think either one is problematic because both new and old options are present in the config file:
```
WARN: The "dns_config.nameservers" configuration key is deprecated. Please use "dns.nameservers.global" instead. "dns_config.nameservers" has been removed.
```

Reply:
see https://github.com/NixOS/nixpkgs/pull/340054#discussion_r1747142639 for the previous thread about this

Reply:
That thread doesn't explain why it would be acceptable to have both `dns` and `dns_config` in config.

Reply:
I agree that this behaviour is not ideal but we can't add assertions for all breaking config changes from upstream. Otherwise we would need to copy all configuration validation logic from upstream which defeats the purpose of RFC42. Before updating users need to read the release notes - both ours and of the software they are using.

Reply:
@fpletz Are you talking about `server_url` and `base_domain`? That's not what this thread is about.

Reply:
> That thread doesn't explain why it would be acceptable to have both `dns` and `dns_config` in config.

I want to emphasize that the module creates both of these, even if the user only specifies `dns`.